Add compression codec configurations for data in exchanges and spills

hackeryang opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->
<!-- Provide an overview of the PR for maintainers and reviewers. -->
## Description
Our community knows that ZSTD compression is better about speed and compression ratio in many scenarios, for example:  https://github.com/trinodb/trino/issues/10058  
Many blogs and test reports on the internet also mentioned that ZSTD is better than ZLIB in most cases:  https://databento.com/blog/zstd-vs-zlib

We found that if page serialization and deserialization also use ZSTD, the **compression and decompression performance** is also better, and `Fault Tolerant Execution` also uses compressed pages:  https://github.com/trinodb/trino/pull/14957

The benchmark result is shown as below, It seems that if compression is enabled, Zstd is about 10X faster than Lz4.  

Because ZSTD may not be a best choice for all scenarios, and we want to take into account the demands of different users, we **added a session property and configuration to switch between several compresion codecs, and still use Lz4 by default** when compression is enabled.
<!-- Provide details that would help an engineer who is unfamiliar with this part of the code. -->
## Additional context and related issues
### Lz4 Compression
```
Benchmark                        (compressed)  (encrypted)  (randomSeed)   Mode  Cnt     Score     Error  Units
BenchmarkPagesSerde.deserialize          true         true          1000  thrpt   10   104.194 ±   4.508  ops/s
BenchmarkPagesSerde.deserialize          true        false          1000  thrpt   10  4770.274 ± 106.577  ops/s
BenchmarkPagesSerde.deserialize         false         true          1000  thrpt   10    83.668 ±   3.172  ops/s
BenchmarkPagesSerde.deserialize         false        false          1000  thrpt   10  8175.870 ± 423.283  ops/s
BenchmarkPagesSerde.serialize            true         true          1000  thrpt   10    90.786 ±  18.355  ops/s
BenchmarkPagesSerde.serialize            true        false          1000  thrpt   10   952.958 ±  31.424  ops/s
BenchmarkPagesSerde.serialize           false         true          1000  thrpt   10    85.247 ±   1.277  ops/s
BenchmarkPagesSerde.serialize           false        false          1000  thrpt   10  8060.584 ± 119.043  ops/s
```
### Zstd Compression
```
Benchmark                        (compressed)  (encrypted)  (randomSeed)   Mode  Cnt      Score      Error  Units
BenchmarkPagesSerde.deserialize          true         true          1000  thrpt   10    102.974 ±    1.746  ops/s
BenchmarkPagesSerde.deserialize          true        false          1000  thrpt   10    373.737 ±   55.635  ops/s
BenchmarkPagesSerde.deserialize         false         true          1000  thrpt   10     85.315 ±    0.897  ops/s
BenchmarkPagesSerde.deserialize         false        false          1000  thrpt   10  10393.729 ± 1296.839  ops/s
BenchmarkPagesSerde.serialize            true         true          1000  thrpt   10     44.524 ±    0.103  ops/s
BenchmarkPagesSerde.serialize            true        false          1000  thrpt   10     65.594 ±    0.373  ops/s
BenchmarkPagesSerde.serialize           false         true          1000  thrpt   10     83.790 ±    2.491  ops/s
BenchmarkPagesSerde.serialize           false        false          1000  thrpt   10   7675.937 ±  347.427  ops/s
```
Benchmarks of exchanges between nodes in a real cluster on AWS, thanks to Mr Raunaq: https://github.com/trinodb/trino/pull/20274#discussion_r1442633339  
Our discussion about when to use ZSTD: https://trinodb.slack.com/archives/CP1MUNEUX/p1704978337669429  
Snappy VS ZSTD:  https://www.percona.com/blog/compression-methods-in-mongodb-snappy-vs-zstd/  

<!-- Mark the appropriate option with an (x). Propose a release note if you can. -->
## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
# General
* The configuration property `exchange.compression-enabled` is now replaced by `exchange.compression-codec` which allows specifying `NONE`, `LZ4` or `ZSTD` as the compression codec for data exchanged between workers. The corresponding system session property `exchange_compression` is replaced by `exchange_compression_codec`. ({issue}`20274`)
* The configuration property `spill-compression-enabled` is now replaced by `spill-compression-codec` which allows specifying `NONE`, `LZ4` or `ZSTD` as the compression codec for data spilled to disk. The already deprecated configuration property `experimental.spill-compression-enabled` is now defunct and must be removed from server configurations. ({issue}`20274`)
```

hackeryang commented:
Hello @findepi @wendigo @losipiuk , can you please take a review when you have time, thank you~

wendigo commented:
I'm concerned that zstd is using jni which could affect gc @hackeryang

wendigo commented:
I don't think that we should switch by default. We could make it configurable

findepi commented:
cc @raunaqmorarka

raunaqmorarka commented:
> I'm concerned that zstd is using jni which could affect gc @hackeryang

This is using the airlift implementation which is pure java code rather than jni unless I'm mistaken about that

raunaqmorarka commented:
> Our community knows that ZSTD compression is better about speed and compression ratio in most scenarios, for example: https://github.com/trinodb/trino/issues/10058

That's a comparison between ZSTD and GZIP, both these prioritise achieving higher compression over CPU efficiency. ZSTD is better between these two, that does not automatically make it the better choice for every scenario compared to lighter compression schemes like SNAPPY and LZ4.
ZSTD makes sense for data files or exchange data stored on S3. But it's not clear that it would be the best choice for data exchanged over the network. With a high speed and bandwidth network we're better off using less CPU heavy compression. The existing LZ4 exchange compression wasn't turned on for pipelined execution because its impact is workload and setup dependent, it makes some queries faster and some slower.
This needs benchmarking in an actual Trino cluster rather than JMH or generic compression/decompression speed tests to assess properly.

wendigo commented:
@raunaqmorarka I've assumed that zstd codec is using zstd-jni underneath, my mistake

hackeryang commented:
> > Our community knows that ZSTD compression is better about speed and compression ratio in most scenarios, for example: #10058
> 
> That's a comparison between ZSTD and GZIP, both these prioritise achieving higher compression over CPU efficiency. ZSTD is better between these two, that does not automatically make it the better choice for every scenario compared to lighter compression schemes like SNAPPY and LZ4. ZSTD makes sense for data files or exchange data stored on S3. But it's not clear that it would be the best choice for data exchanged over the network. With a high speed and bandwidth network we're better off using less CPU heavy compression. The existing LZ4 exchange compression wasn't turned on by default for pipelined execution because its impact is workload and setup dependent, it makes some queries faster and some slower. This needs benchmarking in an actual Trino cluster rather than JMH or generic compression/decompression speed tests to assess properly.

@wendigo @raunaqmorarka you are right, zstd may not be a silver bullet for all situations~ 
I have modified the pr description, and added a configuration with a session property to switch between different codecs, please cc again, it seems that the CICD error is not related to our pr

wendigo commented on core/trino-main/src/main/java/io/trino/FeaturesConfig.java:
```diff
@@ -340,6 +343,19 @@ public FeaturesConfig setExchangeCompressionEnabled(boolean exchangeCompressionE
+        return compressionKind;
+    }
+
+    @Config("compression-kind")
```
"compression-codec"

wendigo commented on core/trino-main/src/main/java/io/trino/FeaturesConfig.java:
```diff
@@ -340,6 +343,19 @@ public FeaturesConfig setExchangeCompressionEnabled(boolean exchangeCompressionE
+    }
+
+    @Config("compression-kind")
+    @ConfigDescription("Compression kind in exchanges, spills and page serializations")
```
"Compression codec used for data in exchanges and spills"

wendigo commented on core/trino-main/src/main/java/io/trino/SystemSessionProperties.java:
```diff
@@ -510,6 +512,12 @@ public SystemSessionProperties(
                         false),
+                enumProperty(
+                        COMPRESSION_KIND,
+                        "Compression kind in exchanges and page serializations",
```
"Compression codec used for data in exchanges and spills"

wendigo replied:
Actually I think that we should have 2 separate configuration properties: one for exchanges (data transmission), one for spills (data storage)

hackeryang replied:
> Actually I think that we should have 2 separate configuration properties: one for exchanges (data transmission), one for spills (data storage)

Yes i haved added a session property and a configuraion for exchanges, and a configuration for spills~ The `FileSingleSpillerFactory` doesn't have a session property api

wendigo commented on core/trino-main/src/main/java/io/trino/execution/buffer/CompressionKind.java:
```diff
@@ -0,0 +1,19 @@
+ */
+package io.trino.execution.buffer;
+
+public enum CompressionKind
```
`CompressionCodec`

wendigo commented on core/trino-main/src/main/java/io/trino/execution/buffer/PagesSerdeFactory.java:
```diff
@@ -27,20 +38,61 @@ public class PagesSerdeFactory
     {
         this.blockEncodingSerde = requireNonNull(blockEncodingSerde, "blockEncodingSerde is null");
         this.compressionEnabled = compressionEnabled;
+        this.compressionKind = compressionKind;
```
`requireNonNull(compressionCodec, "compressionCodec is null")`

wendigo commented on core/trino-main/src/main/java/io/trino/execution/buffer/PagesSerdeFactory.java:
```diff
@@ -27,20 +38,61 @@ public class PagesSerdeFactory
+            case ZSTD:
+                return new ZstdDecompressor();
+        }
+        throw new IllegalArgumentException("Unsupported decompression kind" + compressionKind);
```
missing space between kind and compressionKind

wendigo commented on core/trino-main/src/main/java/io/trino/execution/buffer/PagesSerdeFactory.java:
```diff
@@ -27,20 +38,61 @@ public class PagesSerdeFactory
+            case ZSTD:
+                return new ZstdCompressor();
+        }
+        throw new IllegalArgumentException("Unsupported compression kind" + compressionKind);
```
missing trailing space

wendigo commented on core/trino-main/src/test/java/io/trino/execution/buffer/TestPagesSerde.java:
```diff
@@ -146,8 +148,8 @@ private void testRoundTrip(List<Type> types, List<Page> pages, int blockSizeInBy
         Optional<SecretKey> encryptionKey = encryptionEnabled ? Optional.of(createRandomAesEncryptionKey()) : Optional.empty();
-        PageSerializer serializer = new PageSerializer(blockEncodingSerde, compressionEnabled, encryptionKey, blockSizeInBytes);
-        PageDeserializer deserializer = new PageDeserializer(blockEncodingSerde, compressionEnabled, encryptionKey, blockSizeInBytes);
+        PageSerializer serializer = new PageSerializer(blockEncodingSerde, compressionEnabled, new Lz4Compressor(), encryptionKey, blockSizeInBytes);
```
It'd be nice to test all supported compressors with PageSerializer

hackeryang replied:
> It'd be nice to test all supported compressors with PageSerializer

@wendigo I have added tests about all compressors, they seemed passed, please cc again~

wendigo commented on core/trino-main/src/main/java/io/trino/execution/buffer/PageDeserializer.java:
```diff
@@ -62,14 +61,15 @@ public class PageDeserializer
         encryptionKey.ifPresent(secretKey -> checkArgument(is256BitSecretKeySpec(secretKey), "encryptionKey is expected to be an instance of SecretKeySpec containing a 256bit key"));
         input = new SerializedPageInput(
-                compressionEnabled ? Optional.of(new Lz4Decompressor()) : Optional.empty(),
+                compressionEnabled ? Optional.ofNullable(decompressor) : Optional.empty(),
```
if compression is enabled we should expect non-null decompressor so:

```suggestion
                compressionEnabled ? Optional.of(requireNonNull(decompressor, "decompressor is null")) : Optional.empty(),
```

wendigo commented on core/trino-main/src/main/java/io/trino/execution/buffer/PageSerializer.java:
```diff
@@ -66,14 +63,15 @@ public class PageSerializer
         encryptionKey.ifPresent(secretKey -> checkArgument(is256BitSecretKeySpec(secretKey), "encryptionKey is expected to be an instance of SecretKeySpec containing a 256bit key"));
         output = new SerializedPageOutput(
-                compressionEnabled ? Optional.of(new Lz4Compressor()) : Optional.empty(),
+                compressionEnabled ? Optional.ofNullable(compressor) : Optional.empty(),
```
```suggestion
                compressionEnabled ? Optional.of(requireNonNull(compressor, "compressor is null")) : Optional.empty(),
```

wendigo commented on core/trino-main/src/main/java/io/trino/execution/buffer/PagesSerdeFactory.java:
```diff
@@ -27,20 +38,61 @@ public class PagesSerdeFactory
+    {
+        switch (compressionCodec) {
+            case NONE:
+                return null;
```
I don't like the fact that we are returning null here.

What's the sense of the combination `compressionEnabled == true && compressionCodec == NONE` ?

wendigo replied:
Let's drop NONE and use the `compressionEnabled` information that states explicitly that compression is not used

wendigo replied:
something along these lines: https://github.com/wendigo/trino/commit/c2a7f64b0e870224921a1f45d24eb6fd4bdee93d#diff-6e2a6eb5c7e2eba83c56db26e3868fcc61a0263a479bc052dd67d33ba864d3afR65

hackeryang replied:
> something along these lines: [wendigo@c2a7f64#diff-6e2a6eb5c7e2eba83c56db26e3868fcc61a0263a479bc052dd67d33ba864d3afR65](https://github.com/wendigo/trino/commit/c2a7f64b0e870224921a1f45d24eb6fd4bdee93d#diff-6e2a6eb5c7e2eba83c56db26e3868fcc61a0263a479bc052dd67d33ba864d3afR65)

Thank you very much for the codes~ yes the NONE enum is useless in this case, I just copied it from `io.trino.orc.metadata.CompressionKind` at the beginning~ I have deleted it this time

losipiuk commented on core/trino-main/src/main/java/io/trino/execution/buffer/CompressionCodec.java:
```diff
@@ -0,0 +1,19 @@
+
+public enum CompressionCodec
+{
+    NONE, LZO, SNAPPY, LZ4, ZSTD
```
What is the use of `NONE` if we have separate flag to enable/disable compression

hackeryang replied:
> What is the use of `NONE` if we have separate flag to enable/disable compression

You are right, NONE is useless in this case, I copied the enum from `io.trino.orc.metadata.CompressionKind` at the beginning~ I have deleted the NONE value

losipiuk commented on core/trino-main/src/main/java/io/trino/execution/buffer/PageSerializer.java:
```diff
@@ -94,13 +92,13 @@ private static class SerializedPageOutput
         private static final int INSTANCE_SIZE = instanceSize(SerializedPageOutput.class);
-        // TODO: implement getRetainedSizeInBytes in Lz4Compressor
-        private static final int COMPRESSOR_RETAINED_SIZE = toIntExact(instanceSize(Lz4Compressor.class) + sizeOfIntArray(Lz4RawCompressor.MAX_TABLE_SIZE));
+        // TODO: implement getRetainedSizeInBytes in Lz4Compressor and ZstdCompressor
```
this is prerequisite - otherwise we regress on accuracy of memory management.

hackeryang replied:
> this is prerequisite - otherwise we regress on accuracy of memory management.

I have added back the computation of Lz4RawCompressor.MAX_TABLE_SIZE, thanks for the advice~

losipiuk reviewed (commented):
Nice

raunaqmorarka commented on core/trino-main/src/main/java/io/trino/spiller/NodeSpillConfig.java:
```diff
@@ -14,16 +14,21 @@
     private DataSize queryMaxSpillPerNode = DataSize.of(100, DataSize.Unit.GIGABYTE);
 
     private boolean spillCompressionEnabled;
+    private CompressionCodec spillCompressionCodec = LZ4;
```
Compression codec option for spill to disk should be introduced in a separate commit

raunaqmorarka commented on core/trino-main/src/main/java/io/trino/spiller/NodeSpillConfig.java:
```diff
@@ -14,16 +14,21 @@
     private DataSize maxSpillPerNode = DataSize.of(100, DataSize.Unit.GIGABYTE);
     private DataSize queryMaxSpillPerNode = DataSize.of(100, DataSize.Unit.GIGABYTE);
 
     private boolean spillCompressionEnabled;
```
We should deprecate the boolean flag and use the enum, NONE should be the new replacement for boolean flag. This would consistent with how `io.trino.plugin.hive.HiveCompressionOption` works

hackeryang commented:
@wendigo @losipiuk Thank you for your advice~ I have modified some codes, and the 2 test errors seemed not related to our PR, please cc again~

raunaqmorarka commented on core/trino-main/src/main/java/io/trino/execution/buffer/CompressionCodec.java:
```diff
@@ -0,0 +1,19 @@
+
+public enum CompressionCodec
+{
+    LZO, SNAPPY, LZ4, ZSTD
```
What's the rationale for supporting LZO and SNAPPY here ?
LZ4 offers less aggressive compression at small CPU cost while ZSTD offers aggressive compression at high CPU cost. LZO and SNAPPY are somewhere in between, when would we use either of those two ? We support them in file formats because the standard for those formats requires it. For our own internal usage we should offer fewer choices which fit some use case.

electrum replied:
Agreed. LZ4 should perform better and compress as well as SNAPPY. In other words, LZ4 should be strictly superior to SNAPPY, so there is no need to support it.

raunaqmorarka commented on core/trino-main/src/main/java/io/trino/FeaturesConfig.java:
```diff
@@ -81,6 +83,7 @@ public class FeaturesConfig
      * default value is overwritten for fault tolerant execution in {@link #applyFaultTolerantExecutionDefaults()}}
      */
     private boolean exchangeCompressionEnabled;
```
We should deprecate the boolean flag and instead use NONE enum value to disable compression.

raunaqmorarka commented on core/trino-main/src/main/java/io/trino/FeaturesConfig.java:
```diff
@@ -81,6 +83,7 @@ public class FeaturesConfig
      * default value is overwritten for fault tolerant execution in {@link #applyFaultTolerantExecutionDefaults()}}
      */
     private boolean exchangeCompressionEnabled;
+    private CompressionCodec exchangeCompressionCodec = LZ4;
```
I ran TPCDS sf1000 benchmark on unpartitioned hive tables with ZSTD exchange compression in pipelined execution mode. The setup is a 6 node cluster of r6g.8xlarge machines on AWS.

[TPCDS unpartitioned parquet zstd exchange.pdf](https://github.com/trinodb/trino/files/13839896/TPCDS.unpartitioned.parquet.zstd.exchange.pdf)

<img width="924" alt="Screenshot 2024-01-05 at 2 18 59 PM" src="https://github.com/trinodb/trino/assets/4344846/7294c762-3489-4a91-bf20-de7e98613b5f">

Even though there is a big reduction in internal network data exchanged, there are still regressions in latency due to the high CPU cost of ZSTD compression.
I think this trade-off of more aggressive compression at high CPU cost doesn't make much sense for pipelined execution mode where we exchange data over internal network directly.
This trade off might make sense for filesystem based exchange in fault tolerant execution mode, but that also needs to be assessed with a benchmark.
@losipiuk can we do these changes in way where this config becomes specific to filesystem exchange in FTE ?

raunaqmorarka replied:
Benchmark results with FTE
[FTE unpartitioned hive parquet sf1k zstd compression.pdf](https://github.com/trinodb/trino/files/13870656/FTE.unpartitioned.hive.parquet.sf1k.zstd.compression.pdf)

<img width="1358" alt="Screenshot 2024-01-09 at 1 26 54 PM" src="https://github.com/trinodb/trino/assets/4344846/92f44a03-3937-4329-afdf-3844b6cd5a21">

FTE is using LZ4 exchange compression by default already. Switching to ZSTD improves some queries but it's significantly worse overall. So I'm not seeing a compelling reason to add this option even for FTE.

hackeryang replied:
> Benchmark results with FTE [FTE unpartitioned hive parquet sf1k zstd compression.pdf](https://github.com/trinodb/trino/files/13870656/FTE.unpartitioned.hive.parquet.sf1k.zstd.compression.pdf)
> 
> <img alt="Screenshot 2024-01-09 at 1 26 54 PM" width="1358" src="https://private-user-images.githubusercontent.com/4344846/295137308-92f44a03-3937-4329-afdf-3844b6cd5a21.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQ3OTg3NzUsIm5iZiI6MTcwNDc5ODQ3NSwicGF0aCI6Ii80MzQ0ODQ2LzI5NTEzNzMwOC05MmY0NGEwMy0zOTM3LTQzMjktYWZkZi0zODQ0YjZjZDVhMjEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDEwOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAxMDlUMTEwNzU1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MDQ5N2U3MGMzNDc3NTNhMTVjNzMzODZjNzU2YThlOTUwODcyZWRkNWU2M2NjMzBhYzgyOWNmMGI1NTU1MjhmNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.3PAyEPNHQzIjcY8Jqf1LTD_27ITvoIS30hjP8ETqelE">
> FTE is using LZ4 exchange compression by default already. Switching to ZSTD improves some queries but it's significantly worse overall. So I'm not seeing a compelling reason to add this option even for FTE.

You are right, according to the benchmark results, ZSTD should only be used in clusters with little network bandwidth.

We found that some gaming analytics customers of our company have clusters with strong CPU and poor network cards, maybe in this condition we can use it~  

I saw that `Huawei OpenLookeng`(originated from Trino 350) uses ZSTD in exchanges, but it was for their datacenter connector:  
![image](https://github.com/trinodb/trino/assets/26461591/6879daae-efe8-4a6e-930e-150af2e0bba3)

hackeryang commented:
@raunaqmorarka @electrum Thank you very much for the benchmarks and advice~ I have separated two commits for exchanges and spills, deprecated `boolean compressionEnabled` to use enum `CompressionCodec.NONE`, please review again.

I didn't figure out whether I should modify the serialize size in the test:  
![image](https://github.com/trinodb/trino/assets/26461591/58e3728a-dc63-448a-9ad7-c5697368eead)

hackeryang commented on core/trino-main/src/main/java/io/trino/FeaturesConfig.java:
```diff
@@ -328,18 +331,34 @@ public FeaturesConfig setSpillMaxUsedSpaceThreshold(double spillMaxUsedSpaceThre
+        return exchangeCompressionCodec;
+    }
+
+    @Config("exchange.compression-codec")
```
Thanks for the advice, I have added them in `properties-exchange.md` and `properties-spilling.md`

hackeryang commented:
Relevant slack messages:  https://app.slack.com/client/TFKPX15NC/CP1MUNEUX

hackeryang commented on docs/src/main/sphinx/admin/properties-spilling.md:
```diff
@@ -70,7 +70,15 @@ Limit for memory used for unspilling a single aggregation operator instance.
 - **Default value:** `false`
 
-Enables data compression for pages spilled to disk.
+Enables data compression for pages spilled to disk. It may be replaced by `spill-compression-codec` in future versions.
```
@raunaqmorarka Thanks for the suggestions~ I have modified relevant codes and docs, please cc again  
The CI failures seemed not related to our PR, I saw the former merged PRs also have CI failures

raunaqmorarka commented on core/trino-main/src/main/java/io/trino/execution/buffer/PagesSerdeFactory.java:
```diff
@@ -26,21 +32,49 @@ public class PagesSerdeFactory
+                yield new Lz4Compressor();
+            case ZSTD:
+                yield new ZstdCompressor();
+            default:
```
default branch is unnecessary here

raunaqmorarka commented on core/trino-main/src/main/java/io/trino/execution/buffer/PagesSerdeFactory.java:
```diff
@@ -26,21 +32,49 @@ public class PagesSerdeFactory
+                yield new Lz4Decompressor();
+            case ZSTD:
+                yield new ZstdDecompressor();
+            default:
```
default branch is unnecessary here

hackeryang commented:
It seems CI has been passed(except the irrelevant errors), @raunaqmorarka can you please help merge, thanks~

raunaqmorarka commented:
> It seems CI has been passed(except the irrelevant errors), @raunaqmorarka can you please help merge, thanks~

Please fix the compile error on the first commit

hackeryang commented:
> > It seems CI has been passed(except the irrelevant errors), @raunaqmorarka can you please help merge, thanks~
> 
> Please fix the compile error on the first commit

@raunaqmorarka I have fixed the error, please cc again~

raunaqmorarka commented:
> > > It seems CI has been passed(except the irrelevant errors), @raunaqmorarka can you please help merge, thanks~
> > 
> > 
> > Please fix the compile error on the first commit
> 
> @raunaqmorarka I have fixed the error, please cc again~

We still need a fix to the failing test TestJdbcConnection.testSession

hackeryang commented:
> > > > It seems CI has been passed(except the irrelevant errors), @raunaqmorarka can you please help merge, thanks~
> > > 
> > > 
> > > Please fix the compile error on the first commit
> > 
> > 
> > @raunaqmorarka I have fixed the error, please cc again~
> 
> We still need a fix to the failing test TestJdbcConnection.testSession

@raunaqmorarka Thanks for pointing this out~ I have fixed this test failure, please review again~  

The error this time is about Kudu, seems not related to our PR

wendigo commented on core/trino-main/src/main/java/io/trino/execution/buffer/PagesSerdeFactory.java:
```diff
@@ -26,21 +32,45 @@ public class PagesSerdeFactory
+
+    public static Optional<Compressor> createCompressor(CompressionCodec compressionCodec)
+    {
+        return Optional.ofNullable(switch (compressionCodec) {
+            case NONE:
+                yield null;
+            case LZ4:
+                yield new Lz4Compressor();
+            case ZSTD:
+                yield new ZstdCompressor();
+        });
```
it'd be more readable to return from switch:


```suggestion
        switch (compressionCodec) {
            case NONE:
                return Optional.empty();
            case LZ4:
                return Optional.of(new Lz4Compressor());
            case ZSTD:
                return Optional.of(new ZstdCompressor());
        };
```

wendigo commented on core/trino-main/src/main/java/io/trino/execution/buffer/PagesSerdeFactory.java:
```diff
@@ -26,21 +32,45 @@ public class PagesSerdeFactory
+
+    public static Optional<Decompressor> createDecompressor(CompressionCodec compressionCodec)
+    {
+        return Optional.ofNullable(switch (compressionCodec) {
+            case NONE:
+                yield null;
+            case LZ4:
+                yield new Lz4Decompressor();
+            case ZSTD:
+                yield new ZstdDecompressor();
+        });
```
```suggestion
        switch (compressionCodec) {
            case NONE:
                return Optional.empty();
            case LZ4:
                return Optional.of(new Lz4Decompressor());
            case ZSTD:
                return Optional.of(new ZstdDecompressor());
        });
```

hackeryang replied:
@wendigo Thanks for the advice~ I have modified relevant codes, please cc again.

Mr raunaq wanted to remove the `case default:` branch, but it will cause a syntax error if we use `switch (compressionCodec)`, so I finally used the `return switch()` syntax and lambda expressions suggested by Intellij IDEA~

wendigo replied:
Looks good now :)

wendigo reviewed (approved):
LGTM % two comments
